### PR TITLE
S39: Endpoints for CRUD tickets

### PR DESCRIPTION
**Pull request by qin-guan**
- Updated Dtos namespace to DTOs
- Added update ticket endpoint
- Added JsonPatch
- Updated project update endpoint
- Renamed task to ticket in permissions
- Removed unused converters

### CODE REVIEW

File: Controllers/TicketsController.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                    await _authorizationService.AuthorizeAsync(User, project, Permission.CreateTicket.ToString());
+                if (!authorization.Succeeded) return Forbid();
+                
+                var ticket = _mapper.Map<Ticket>(createTicketRequestDto);
+                if (ticket.ParentTicketId is not null &&
+                    await _dbContext.Tickets.FindAsync(ticket.ParentTicketId) is null) return NotFound();
+                if (ticket.ReporterId is not null && project.Users.SingleOrDefault(u => u.Id == ticket.ReporterId) ==
+                    default(ApplicationUser)) return BadRequest();
+                if (ticket.AssigneeId is not null && project.Users.SingleOrDefault(u => u.Id == ticket.AssigneeId) ==
+                    default(ApplicationUser)) return BadRequest();
```
Comment (julwrites) [inline Controllers/TicketsController.cs:79]: I'm a bit confused as to how to read this; is it that `ReporterId` and `AssigneeId`, if they exist, must point to default `ApplicationUser`?

File: DTOs/Ticket/CreateTicketRequestDto.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        [Required] public string Name { get; set; }
+        public string Description { get; set; }
+
+        [Required]
+        [EnumDataType(typeof(TicketType))]
+        public TicketType TicketType { get; set; }
+
+        [Required]
+        [EnumDataType(typeof(TicketStatus))]
+        public TicketStatus TicketStatus { get; set; }
```
Comment (julwrites) [inline DTOs/Ticket/CreateTicketRequestDto.cs:19]: Do we need to think about extensibility, or will this be easy to extend later (e.g. adding labels, links, attachments, etc...)

File: Models/Permission.cs [hunk 3]
```diff
@@ -6,8 +6,8 @@ public enum Permission
         UpdateProject,
         DeleteProject,
         
-        CreateTask,
-        UpdateTask,
+        CreateTicket,
+        UpdateTicket,
         DeleteTask,
```
Comment (julwrites) [inline Models/Permission.cs]: `DeleteTicket`?

File: Models/Project.cs [hunk 4]
```diff
@@ -9,6 +9,7 @@ public class Project
         public string Name { get; set; }
         public string Description { get; set; }
         public Uri Repository { get; set; }
+        public List<Ticket> TaskItems { get; set; }
```
Comment (julwrites) [inline Models/Project.cs]: `Tickets`?

File: Models/Ticket.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        public Guid Id { get; set; }
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public TicketType TicketType { get; set; }
+        public TicketStatus TicketStatus { get; set; }
+        public Guid ProjectId { get; set; }
+        public Project Project { get; set; }
+        public Guid? ReporterId { get; set; }
+        public virtual ApplicationUser Reporter { get; set; }
```
Comment (julwrites) [inline Models/Ticket.cs:15]: Why have both the `Reporter` and `ReporterId`?

**Review by julwrites — COMMENTED** [review]
If we want to rename `Task` to `Ticket`, we should do it everywhere.

File: Controllers/TicketsController.cs (hunk 1 above)
Comment (qin-guan) [reply Controllers/TicketsController.cs:79]: If a Reporter/Assignee is provided they must exist inside the project, if the default of `ApplicationUser` is returned, it means they do not exist

File: DTOs/Ticket/CreateTicketRequestDto.cs (hunk 2 above)
Comment (qin-guan) [reply DTOs/Ticket/CreateTicketRequestDto.cs:19]: The DTO would need to be updated then, but it shouldn't be too hard...

File: Models/Permission.cs (hunk 3 above)
Comment (qin-guan) [reply Models/Permission.cs]: Ah...

File: Models/Ticket.cs (hunk 5 above)
Comment (qin-guan) [reply Models/Ticket.cs:15]: Because when creating the ticket it is easier to set the Id

File: Controllers/TicketsController.cs (hunk 1 above)
Comment (julwrites) [reply Controllers/TicketsController.cs:79]: Ohhhhh okay right.

**Review by julwrites — APPROVED** [review]

**Pull request merged**